### PR TITLE
Fix Linux and MacOS file permissions:

### PR DESCRIPTION
**Pull request by PJayB**
# Description:

Fix Linux and MacOS file permissions:

* `SetExecutable`: Today, existing permissions are destroyed. The fix preservss any existing read-write permissions and only adds `x`.
* `SetReadOnly`: Generally, a file should only be writable for the user that created it. Thus, when removing the flag, remove `w` from everyone. When adding the flag, only set for the user.
* `mkdir`: Create directories with 0755 permissions. By default, directories shouldn't be writable by anyone but the user that created it.

Fixes issue https://github.com/fastbuild/fastbuild/issues/200.

# Checklist:

The pull request:
- [X] **Is created against the Dev branch**
- [X] **Is self-contained**
- [X] **Compiles on Windows, OSX and Linux** (TODO - will test on Linux & Windows, but will need help with MacOS)
- [X] **Has accompanying tests** (I figured the existing coverage would be sufficient, but please let me know if you want to see more detailed tests)
- [X] **Passes existing tests**
- [X] **Keeps Windows, OSX and Linux at parity**
- [X] **Follows the code style**
- [X] **Includes documentation** (fixes for existing feature)

### CODE REVIEW

File: Code/Core/FileIO/FileIO.cpp [hunk 1]
```diff
@@ -898,12 +898,12 @@
         if ( readOnly )
         {
             // remove writable flag
-            s.st_mode &= ( ~S_IWUSR ); // TODO:LINUX Is this the correct behaviour?
+            s.st_mode &= ~( S_IWUSR | S_IWGRP | S_IWOTH );
```
Comment (ffulin) [inline Code/Core/FileIO/FileIO.cpp:901]: If setting writeable only adds the user flag, it seems that removing writable should also only remove thwe user writable flag?

**Comment by ffulin** [comment]
Thanks for taking a look at this. I think these changes make sense, with the exception of SetReadOnly behavior not being symmetrical which doesn't seem correct to me.

**Comment by PJayB** [comment]
If you want to think of it symmetrically, one way to think about it is that SetReadOnly is doing the most conservative operation in either direction.

When marking something as readonly, it wouldn't make sense to only mark the file as RO for just the User that created it, but leave write available to the wider Group or Other. Doing anything else invalidates the purpose.

Conversely, when setting write permissions, typically only the User has write permissions unless that user explicitly grants Group or Other access. So we only want to set the writeable bit in this case.

**Comment by PJayB** [comment]
@ffulin I've incorporated some MacOS build fixes from @Leandros. He confirmed over on #200 that FBuild builds and tests pass on MacOS with this change.

Does the above comment about the read-only bits make sense?

File: Code/Core/Env/Assert.cpp [hunk 2]
```diff
@@ -85,7 +85,7 @@ bool IsDebuggerAttached()
         const uint32_t BUFFER_SIZE( 4096 );
         char buffer[ BUFFER_SIZE ];
         #if defined( __APPLE__ ) || defined( __LINUX__ )
-            sprintf( buffer,
+            snprintf( buffer, BUFFER_SIZE,
```
Comment (ffulin) [inline Code/Core/Env/Assert.cpp:88]: These snprintf changes seem ok but unrelated to file permissions?

Comment (PJayB) [reply Code/Core/Env/Assert.cpp]: Removed. @Leandros needed these changes to build for OSX, but I'll let you guys work that out.

**Comment by ffulin** [comment]
Thanks for the fixes! These have been integrated and will be in the next release (v1.10)
